cli/util/uploader.go

```diff
@@ -0,0 +1,352 @@
+		return 0, finalErr
+	}
+
+	return finalResult, nil
+}
+
+func fetchBlockData(rpcClient *rpcclient.Client, index uint) ([]byte, error) {
+	if index%1000 == 0 {
+		fmt.Println("Fetching block", index)
+	}
+	block, err := rpcClient.GetBlockByIndex(uint32(index))
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch block %d: %w", index, err)
+	}
+
+	bw := io.NewBufBinWriter()
+	block.EncodeBinary(bw.BinWriter)
+
+	if bw.Err != nil {
+		return nil, fmt.Errorf("failed to encode block %d: %w", index, bw.Err)
+	}
+
+	return bw.Bytes(), nil
+}
+
+func uploadBObjWithSlicer(ctx context.Context, clientSDK *client.Client, account wallet.Account, containerID cid.ID, objData []byte, attributeKey, attributeValue string, sessionToken *session.Object) error {
+	signer := user.NewAutoIDSignerRFC6979(account.PrivateKey().PrivateKey)
+	var ownerID user.ID
+	ownerID.SetScriptHash(account.PrivateKey().GetScriptHash())
+
+	slc, err := slicer.New(ctx, clientSDK, signer, containerID, ownerID, sessionToken)
+	if err != nil {
+		return fmt.Errorf("failed to create slicer: %w", err)
+	}
+
+	attrs := []object.Attribute{
+		*object.NewAttribute(attributeKey, attributeValue),
+	}
+
+	_, err = slc.Put(ctx, bytes.NewReader(objData), attrs)
+	if err != nil {
+		return fmt.Errorf("failed to slice and upload block data: %w", err)
+	}
+
+	return nil
+}
+
+// updateIndexFiles updates the index files in the container.
+func updateIndexFiles(ctx *cli.Context, clientSDK *client.Client, containerID cid.ID, account wallet.Account, currentHeight uint, attributeKey string, blockAttributeKey string) error {
+	ctx.App.Writer.Write([]byte("Updating index files...\n"))
+
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 0), object.MatchNumGE)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, clientSDK, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("no OIDs found for index files: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / uint(indexFileSize)
+
+	if existingIndexCount == expectedIndexCount {
+		return nil
+	}
+
+	var buf bytes.Buffer
+	fileIndex := int(existingIndexCount)
+
+	ctxWithCancel, cancel := context.WithCancel(ctx.Context)
+	defer cancel()
+
+	workerPool := make(chan struct{}, maxParallelSearches)
+	errCh := make(chan error, 1)
+	wg := sync.WaitGroup{}
+
+	for i := existingIndexCount; i < expectedIndexCount; i++ {
+		startIndex := i * indexFileSize
+		endIndex := startIndex + indexFileSize
+		if endIndex > currentHeight {
+			ctx.App.Writer.Write([]byte(fmt.Sprintf("end index %d exceeds current height %d", endIndex, currentHeight)))
+			return nil
+		}
+
+		blockOids := make([]oid.ID, indexFileSize)
+		for j := startIndex; j < endIndex; j++ {
+			select {
+			case <-ctxWithCancel.Done():
+				return <-errCh
+			default:
+			}
+
+			wg.Add(1)
+			workerPool <- struct{}{}
```

Comment:
It should not be needed, this code needs to be refactored.

Reply:
Do you mean fully remove concurrency?  The consequent execution even with the `getHeader` approach will take 4 hours for one index file generation (128k).

Reply:
Discussed, concurrency is OK, but the code itself needs to be refactored (workerPool), I left a separate comment about it.